### PR TITLE
Edit and delete

### PR DESCRIPTION
**Pull request by Smital-Lunawat**
Delete is working fine, edit needs to be called on the backend. It is not making correct calls.

### CODE REVIEW

**Comment by marcosfrenkel** [comment]
Dialog should take more space

**Review by marcosfrenkel — CHANGES_REQUESTED** [review]

File: dragon-scales/src/app/components/TaskViewerComponents/TaskViewer.jsx
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/TaskViewerComponents/TaskViewer.jsx:1]: Something here is crashing when you try and edit a content block

File: dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx [hunk 1]
```diff
@@ -47,107 +56,153 @@ const StyledStepPaperActive = styled(Paper)(({ theme }) => ({
...
+    paddingTop: "10px",
+}));
 
+function StepViewer({ 
+    stepEntity, 
+    markStepState = () => {}, 
+    reloadTask = () => {}, 
+    parentID = '', 
+    parentName = '', 
+}) {
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx:82]: why are you changing the signature here? None of these arguments should have defaults, they need a reload task and markStepState argument there.

File: dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx [hunk 2]
```diff
@@ -47,107 +56,153 @@ const StyledStepPaperActive = styled(Paper)(({ theme }) => ({
...
     const { entitySectionIdRef } = useContext(ExplorerContext);
 
+    // Holds as keys the ids of contentBlocks and as values their corresponding refs
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx:85]: That is not the line where the comment was, this is specifically for contentBlocksRefs

File: dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx [hunk 3]
```diff
@@ -47,107 +56,153 @@ const StyledStepPaperActive = styled(Paper)(({ theme }) => ({
...
-
-
-    // Holds as keys the ids of contentBlocks and as values their corresponding refs
+    const [editDialogOpen, setEditDialogOpen] = useState(false);
+    
     const contentBlocksRefs = useRef({});
     const activeContentBlockRef = useRef(null);
-    const newContentBlockRef = useRef(null);  // Used to keep track of the up-to-date content of tiptap
-    const stepViewerRef = useRef(null); // used to keep track of active state
+    const newContentBlockRef = useRef(null);
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx:95]: Why are these comments missing?

File: dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx [hunk 4]
```diff
@@ -47,107 +56,153 @@ const StyledStepPaperActive = styled(Paper)(({ theme }) => ({
...
+    const stepViewerRef = useRef(null);
     const stepRef = useRef(null);
 
     entitySectionIdRef.current[step.ID] = stepRef;
 
+    // Update step when stepEntity changes
+    useEffect(() => {
+        setStep(stepEntity);
+    }, [stepEntity]);
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx:104]: When does this happen? do you ever touch stepEntity after creation?

File: dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx [hunk 5]
```diff
@@ -47,107 +56,153 @@ const StyledStepPaperActive = styled(Paper)(({ theme }) => ({
...
+    useEffect(() => {
+        if (step?.comments) {
+            const parsedBlocks = step.comments.map(comment => JSON.parse(comment));
+            setParsedContentBlocksEnt(parsedBlocks);
+            contentBlocksRefs.current = parsedBlocks.reduce((acc, block) => {
+                acc[block.ID] = block.content[block.content.length - 1];
+                return acc;
+            }, {});
+        }
+    }, [step]);
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx:116]: This useEffect is the same as the one removed at the bottom but removed the FIXME message? why did you remove the FIXME?

File: dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx [hunk 6]
```diff
@@ -47,107 +56,153 @@ const StyledStepPaperActive = styled(Paper)(({ theme }) => ({
...
+
+    // Update entitySectionIdRef when step ID changes
+    useEffect(() => {
+        if (step?.ID && entitySectionIdRef?.current) {
+            entitySectionIdRef.current[step.ID] = stepRef;
+        }
+    }, [step?.ID, entitySectionIdRef]);
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx:123]: when does this happen? the id of an entity never changes right?

File: dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx [hunk 7]
```diff
@@ -281,7 +315,41 @@ export default function StepViewer( { stepEntity, markStepState, reloadTask } )
...
+                open={editDialogOpen}
+                onClose={handleCloseEditDialog}
+                reloadParent={reloadTask}
+                onEditSuccess={handleEditSuccess}
+            />
+        </Box>
+    );
+}
+
+StepViewer.propTypes = {
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx:343]: Can you explain a little how this work? seems like a typehint stuff

File: dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx (hunk 1 above)
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx:82]: Also, parent ID is in the stepEntity, keep a single source of truth.

File: dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx [hunk 8]
```diff
@@ -47,107 +56,153 @@ const StyledStepPaperActive = styled(Paper)(({ theme }) => ({
...
+    };
 
-    const deactivateStepViewer = () => {
+    const deactivateStepViewer = async () => {
         setIsActive(false);
-        markStepState(step.ID, false);
+        try {
+            markStepState(step.ID, false);
+        } catch (error) {
+            console.error('Error in markStepState:', error);
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx:163]: what error are you catching?

File: dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx [hunk 9]
```diff
@@ -47,107 +56,153 @@ const StyledStepPaperActive = styled(Paper)(({ theme }) => ({
...
-            ).then(response => {
-                    if (success) {
-                        activeContentBlockRef.current = null;
-                        getEntity(step.ID).then(entity => {
-                            setStep(JSON.parse(entity));
-                        })
-                    } else {
-                        console.log("Error: content block edition failed")
-                    }
+            try {
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx:170]: Why are FIXME messages gone here?

File: dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx [hunk 10]
```diff
@@ -169,108 +224,87 @@ export default function StepViewer( { stepEntity, markStepState, reloadTask } )
...
+    };
+
+    const handleCloseEditDialog = () => {
+        setEditDialogOpen(false);
+    };
 
+    // Handle successful edit
+    const handleEditSuccess = (updatedStep) => {
+        setStep(updatedStep);
+    };
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/StepViewerComponents/StepViewer.jsx:240]: What is this doing? If I understand this correctly, on success you are just triggering the reload function right? so why are you changing step here and what is updatedStep? I can't get a log statement on my browser when changing the step name

File: dragon-scales/src/app/components/dialogs/EditEntityDialog.jsx [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    );
+}
+
+export default function EditEntityDialog({ 
+    user, 
+    type, 
+    entityName, 
+    entityID,
+    parentID,
+    parentName,
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/dialogs/EditEntityDialog.jsx:34]: why do you need parent ID?

File: dragon-scales/src/app/components/TaskViewerComponents/TaskViewer.jsx [hunk 12]
```diff
@@ -1,223 +1,208 @@
...
 import { ExplorerContext } from "@/app/contexts/explorerContext";
 
-const StyledDeleteButton = styled(IconButton)(({ theme }) => ({
-    position: 'absolute',
-    top: theme.spacing(2),
-    right: theme.spacing(2),
-    color: theme.palette.error.main,
-}));
-
+// Styled components remain the same...
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/TaskViewerComponents/TaskViewer.jsx:20]: This feels like a comment for the PR not the code, remove this one.

File: dragon-scales/src/app/components/TaskViewerComponents/TaskViewer.jsx [hunk 13]
```diff
@@ -1,223 +1,208 @@
...
+                console.error("Error submitting content block.");
+            }
         }
-    }
-
-    const updateStepActiveStatus = (stepId, isActive) => {
-        setActiveSteps(prevState => ({
-            ...prevState,
-            [stepId]: isActive
-        }));
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/TaskViewerComponents/TaskViewer.jsx:98]: Why is this gone? doesn't the task need to know what step is active?

File: dragon-scales/src/app/components/TaskViewerComponents/TaskViewer.jsx [hunk 14]
```diff
@@ -1,223 +1,208 @@
...
+                <form noValidate autoComplete="off" onSubmit={handleSubmitNewContent}>
+                    <StyledNewContentBox>
+                        <ViewCompactIcon />
+                        <Tiptap
+                            onContentChange={handleNewContentBlockChange}
+                            entID={task.ID}
+                            initialContent={newContentBlockRef.current}
+                            reloadEditor={reloadEditor}
+                            placeholder={`Add content block to "${task.name}" here...`}
+                            newLineEditor
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/TaskViewerComponents/TaskViewer.jsx:191]: Where did the true go here?

File: dragon-scales/src/app/components/ExplorerDrawer.jsx [hunk 15]
```diff
@@ -1,9 +1,9 @@
 "use client";
 
-// TODO: Only 1 thing across all of the trees should be selected at a time. If an item is not selected and the user clicks it once, it should not expand but instead scroll there
+// TODO: Ensure only one item is selected at a time across all trees.
```
Comment (marcosfrenkel) [inline dragon-scales/src/app/components/ExplorerDrawer.jsx:3]: Where did the second part of that fixme go?

**Review by marcosfrenkel — CHANGES_REQUESTED** [review]

File: dragon-scales/src/app/components/TaskViewerComponents/TaskViewer.jsx
Comment (Smital-Lunawat) [reply dragon-scales/src/app/components/TaskViewerComponents/TaskViewer.jsx:1]: I am not able to resolve this can you please help me figure this out

File: dragon-scales/src/app/components/dialogs/EditEntityDialog.jsx (hunk 11 above)
Comment (Smital-Lunawat) [reply dragon-scales/src/app/components/dialogs/EditEntityDialog.jsx:34]: No it was not needed, removed it now

File: dragon-scales/src/app/components/TaskViewerComponents/TaskViewer.jsx (hunk 14 above)
Comment (Smital-Lunawat) [reply dragon-scales/src/app/components/TaskViewerComponents/TaskViewer.jsx:191]: It behaves similarly right now without the true, if we remove the line entirely then the interface design looks bad, thats it.
